react/features/connection-indicator/components/AbstractConnectionIndicator.js

```diff
@@ -0,0 +1,179 @@
+     * returns {void}
+     */
+    componentDidUpdate(prevProps: Props) {
+        if (prevProps.participantId !== this.props.participantId) {
+            statsEmitter.unsubscribeToClientStats(
+                prevProps.participantId, this._onStatsUpdated);
+            statsEmitter.subscribeToClientStats(
+                this.props.participantId, this._onStatsUpdated);
+        }
+    }
+
+    /**
+     * Cleans up any queued processes, which includes listening for new stats
+     * and clearing any timeout to hide the indicator.
+     *
+     * @private
+     * @returns {void}
+     */
+    componentWillUnmount() {
+        statsEmitter.unsubscribeToClientStats(
+            this.props.participantId, this._onStatsUpdated);
+
+        if (this.state.autoHideTimeout) {
+            clearTimeout(this.state.autoHideTimeout);
+        }
+    }
+
+    _onStatsUpdated: (Object) => void;
+
+    /**
+     * Callback invoked when new connection stats associated with the passed in
+     * user ID are available. Will update the component's display of current
+     * statistics.
+     *
+     * @param {Object} stats - Connection stats from the library.
+     * @private
+     * @returns {void}
+     */
+    _onStatsUpdated(stats = {}) {
+        // Rely on React to batch setState actions.
+        const { connectionQuality } = stats;
+        const newPercentageState = typeof connectionQuality === 'undefined'
+            ? {} : { percent: connectionQuality };
+        const newStats = Object.assign(
+            {},
+            this.state.stats,
+            stats,
+            newPercentageState);
+
+        this.setState({
+            stats: newStats
+        });
+
+        this._updateIndicatorAutoHide(newStats.percent);
+    }
+
+    /**
+     * Updates the internal state for automatically hiding the indicator.
+     *
+     * @param {number} percent - The current connection quality percentage
+     * between the values 0 and 100.
+     * @private
+     * @returns {void}
+     */
+    _updateIndicatorAutoHide(percent) {
+        if (percent < INDICATOR_DISPLAY_THRESHOLD) {
+            if (this.state.autoHideTimeout) {
+                clearTimeout(this.state.autoHideTimeout);
+            }
+
+            this.setState({
+                autoHideTimeout: undefined,
+                showIndicator: true
+            });
+        } else if (this.state.autoHideTimeout) {
+            // This clause is intentionally left blank because no further action
+            // is needed if the percent is below the threshold and there is an
+            // autoHideTimeout set.
+        } else {
+            this.setState({
+                autoHideTimeout: setTimeout(() => {
+                    this.setState({
+                        showIndicator: false
+                    });
+                }, typeof interfaceConfig === 'undefined'
+                    ? 5000
+                    : interfaceConfig.CONNECTION_INDICATOR_AUTO_HIDE_TIMEOUT)
```

Comment:
Maybe conceptually this value is better as a prop.

Reply:
I'd actually even remove this. Can't imagine a situation where someone would feel a strong urge to configure this. Especially as some other "hide timeout"s are not configurable in the app. What do you think?

Reply:
IIRC there was an option to completely hide it. Me likey because it makes the UI cleaner and some users are confused adn want to disable it.